feat(date): allow restrictions for dates and query date ranges

Chartman123 opened the pull request:
This closes  #2396 by allowing to set minimum and maximum dates and to query date ranges

Set question settings for minimum/maximum dates and if the user must enter a date range:
![image](https://github.com/user-attachments/assets/ffb2555f-9c5d-4ac8-8066-0c6dc6d8bd50)

Restricted dates and date range picker in submit view:
![grafik](https://github.com/user-attachments/assets/997daac0-5c9e-4002-a348-1cfb6bc2d15b)

Results in summary / responses view:
![grafik](https://github.com/user-attachments/assets/291b1ff1-cfba-4949-af7f-633518e63689) 
![grafik](https://github.com/user-attachments/assets/fcdb4042-8797-4605-b841-2ffcbf7b0a2e)

Signed-off-by: Christian Hartmann <chris-hartmann@gmx.de>

Chartman123 commented:
Currently the Calendar of the NcDateTimePicker is not shown correctly in the menu:

![image](https://github.com/user-attachments/assets/2667f40f-d795-40dd-93b6-dd2b65acece7)

@susnux is this a bug in the `@nextcloud/vue` lib?

jancborchardt commented on src/components/Questions/QuestionDate.vue:
```diff
@@ -9,6 +9,37 @@
+			<NcActionInput
+				v-model="dateMin"
+				type="date"
+				:label="t('forms', 'Pick minimum date')"
```
```suggestion
				:label="t('forms', 'Earliest date')"
```

jancborchardt commented on src/components/Questions/QuestionDate.vue:
```diff
@@ -9,6 +9,37 @@
+			<NcActionCheckbox
+				:checked="extraSettings?.dateRange"
+				@update:checked="onDateRangeChange">
+				{{ t('forms', 'Query date range') }}
```
```suggestion
				{{ t('forms', 'Limit date range') }}
```

Chartman123 replied:
I changed the label for the checkbox to "Use date range" to make it clearer

jancborchardt commented on src/components/Questions/QuestionDate.vue:
```diff
@@ -9,6 +9,37 @@
+			<NcActionInput
+				v-model="dateMax"
+				type="date"
+				:label="t('forms', 'Pick maximum date')"
```
```suggestion
				:label="t('forms', 'Latest date')"
```

jancborchardt commented on src/components/Questions/QuestionDate.vue:
```diff
@@ -9,6 +9,37 @@
+				:formatter="extraSettingsFormatter"
+				is-native-picker
+				:max="dateMax">
+				<template #icon>
+					<Pencil :size="20" />
+				</template>
```
```suggestion
```

Chartman123 replied:
See comment below

jancborchardt commented on src/components/Questions/QuestionDate.vue:
```diff
@@ -9,6 +9,37 @@
+				:formatter="extraSettingsFormatter"
+				is-native-picker
+				:min="dateMin">
+				<template #icon>
+					<Pencil :size="20" />
+				</template>
```
```suggestion
```

Chartman123 replied:
See comment below

jancborchardt reviewed (changes requested):
Looks nice! :) Added some wording suggestions to make it a bit more easy to understand / less technical.
And I’d say the icons can be removed as they are the same and thus mostly visual noise? It’s a sub-item of the checkbox anyway so it’s fine to not have an icon and thus look a bit indented.

Chartman123 commented:
@jancborchardt thanks for the review :) 

> It’s a sub-item of the checkbox anyway

No, they are separate from the checkbox and always shown. The restrictions of which date can be choses apply also to single date questions. That's why I added the icons. Without them it felt wrong to me.

jancborchardt reviewed (commented):
As icons, how about:
- "Today" for Earliest date
- "Event" for Latest date

https://fonts.google.com/icons?icon.set=Material+Symbols&icon.size=24&icon.color=%235f6368&icon.query=event

Chartman123 commented:
> As icons, how about:
> 
> * "Today" for Earliest date
> * "Event" for Latest date
> 
> https://fonts.google.com/icons?icon.set=Material+Symbols&icon.size=24&icon.color=%235f6368&icon.query=event

I've changed the icons and updated the screenshot in the OP :)

susnux commented on openapi.json:
```diff
@@ -418,6 +418,17 @@
+                        "type": "integer",
+                        "format": "int64"
+                    },
+                    "dateRange": {
```
Maye it makes sense to have `ANSWER_TYPE_DATE` <-> `ANSWER_TYPE_DATE_RANGE`, `ANSWER_TYPE_DATETIME` <->`ANSWER_TYPE_DATETIME_RANGE` ?

But also fine with this.

jancborchardt replied:
@susnux how would that look UI-wise, or what would be the choices in plain English? :D

Chartman123 replied:
As our UI elements support the range and non-range function in one and also only a few modifications were needed, I think we can leave it this way :) and I think it doesn't make much difference if we query the question.type vs. the extraSetting. Should I include the same for time questions into this PR or a follow-up?

Chartman123 commented:
> sorry, for some reason the icon link didn’t save the "fill" attribute. :\ The icons should use the filled versions, as we always do.

@jancborchardt fixed :)

jancborchardt reviewed (approved):
Really nice! :)

susnux reviewed (approved):
I think this is simpler than 4 date types.

Bibo-Joshi commented:
This is awesome, thanks a ton! 🎊
